neps/nep-0393.md

```diff
@@ -0,0 +1,596 @@
+
+    /// Returns total amount of tokens of given class minted by `issuer`. See `sbt_supply` for
+    /// information about revoked tokens.
+    fn sbt_supply_by_class(&self, issuer: AccountId, class: ClassId) -> u64;
+
+    /// Returns total supply of SBTs for a given owner. See `sbt_supply` for information about
+    /// revoked tokens.
+    /// If `class` is specified, returns only owner supply of the given class (either 0 or 1).
+    fn sbt_supply_by_owner(
+        &self,
+        account: AccountId,
+        issuer: AccountId,
+        class: Option<ClassId>,
+    ) -> u64;
+
+    /// Query sbt tokens issued by a given contract.
+    /// `limit` specifies the upper limit of how many tokens we want to return.
+    /// If `from_token` is not specified, then `from_token` should be assumed
+    /// to be the first valid token id. If `with_expired` is set to `true` then all the tokens are returned
+    /// including expired ones otherwise only non-expired tokens are returned.
+    fn sbt_tokens(
+        &self,
+        issuer: AccountId,
+        from_token: Option<u64>,
+        limit: Option<u32>,
+        with_expired: bool,
+    ) -> Vec<Token>;
+
+    /// Query SBT tokens by owner.
+    /// `limit` specifies the upper limit of how many tokens we want to return.
+    /// If `from_class` is not specified, then `from_class` should be assumed to be the first
+    /// valid class id. If `with_expired` is set to `true` then all the tokens are returned
+    /// including expired ones otherwise only non-expired tokens are returned.
+    /// Returns list of pairs: `(Contract address, list of token IDs)`.
+    fn sbt_tokens_by_owner(
+        &self,
+        account: AccountId,
+        issuer: Option<AccountId>,
+        from_class: Option<u64>,
+        limit: Option<u32>,
+        with_expired: bool,
+    ) -> Vec<(AccountId, Vec<OwnedToken>)>;
+
+    /// checks if an `account` was banned by the registry.
+    fn is_banned(&self, account: AccountId) -> bool;
+
+    /*************
+     * Transactions
+     *************/
+
+    /// Creates a new, unique token and assigns it to the `receiver`.
+    /// `token_spec` is a vector of pairs: owner AccountId and TokenMetadata.
+    /// Each TokenMetadata must have non zero `class`.
+    /// Must be called by an SBT contract.
+    /// Must emit `Mint` event.
+    /// Must provide enough NEAR to cover registry storage cost.
+    // #[payable]
+    fn sbt_mint(&mut self, token_spec: Vec<(AccountId, Vec<TokenMetadata>)>) -> Vec<TokenId>;
+
+    /// sbt_recover reassigns all tokens issued by the caller, from the old owner to a new one.
+    /// Adds `old_owner` to a banned accounts list.
+    /// Must not add `old_owner` to a ban list.
+    /// Must be called by a valid SBT issuer.
+    /// Must emit `Recover` event.
+    /// Requires attaching enough NEAR to cover the storage growth.
+    // #[payable]
+    fn sbt_recover(&mut self, from: AccountId, to: AccountId) -> (u32, bool);
```

Comment:
I totally missed the return value here in the initial review.

`[1, true]`. Go figure what it means, and it is also not extendable. I recommend returning an object with explicit field names and avoid boolean fields, instead, return enum `"status": "FAILED" | "IN_PROGRESS" | "SUCCEEDED"` (I don't know what that bool signals in the first place)

Reply:
@robert-zaremba I am not suggesting nullability, I suggest explicitly named fields and values, so return:

```json
{
  "status": "NEEDS_FURTHER_RECOVERY",
  "number_of_sbts_left": 10
}
```

instead of:

```json
[10, true]
```

The current type is not self-descriptive.